Enhance shape inference in ONNXIFI transformer

yinghai opened the pull request:
Summary:
In this diff, we push the fake run of the net into the ONNXIFI transformer, because
1. We cannot do shape inference for every op
2. Since the net has been SSA rewritten, we cannot use shape info from outer workspace directly.

In addition, this diff adds input shape info when querying the `onnxBackendCompatibility` function.

Differential Revision: D10390164

bddppq commented on caffe2/python/pybind_state.cc:
```diff
@@ -1673,6 +1673,7 @@ void addGlobalMethods(py::module& m) {
       "onnxifi",
       [](const py::bytes& pred_net_str,
          const std::unordered_map<std::string, std::vector<int>>& shapes,
+         bool infer_shapes,
```
what's the reason of not always doing shape inference?

yinghai replied:
Sometimes shape inference is not feasible as the output shape depends not only the input shape but also the input value. One example is those SparseNN ops. In this case, we have to do a fake run. 

Another reason is that our shape inference is not very robust. It might fail even it's not the case above.

bddppq replied:
I see. I think the first reason shouldn't prevent us to always set infer_shapes to true? (while the second one is the real blocker).

yinghai replied:
Actually I think the second one is the real blocker. If the shape inference is buggy we can still fix. But if there is no way to infer shape just from the input shape, we have to do a RunNet() to populate the shapes.

bddppq commented on caffe2/python/onnx/test_onnxifi.py:
```diff
@@ -212,7 +212,8 @@ def test_resnet50_core(self):
         pred_net_cut = onnxifi_caffe2_net(pred_net,
-                                          {input_name: input_blob_dims})
+                                          {input_name: input_blob_dims},
+                                          True)
```
Maybe use `infer_shapes=True`?

bddppq commented on caffe2/opt/backend_cutting.cc:
```diff
@@ -238,7 +251,12 @@ void DetectBoundaryReferences(
     } else {
       for (auto child_node : nn::getConsumers(node)) {
-        const auto& info = infos.at(child_node);
+        auto child_it = infos.find(child_node);
```
maybe create a function to avoid the duplication?

bddppq commented on caffe2/opt/onnxifi_transformer.cc:
```diff
@@ -84,7 +97,8 @@ void FillModelInfo(::ONNX_NAMESPACE::ModelProto* model) {
 } // namespace
 
-OnnxifiTransformer::OnnxifiTransformer(bool debug) : debug_(debug) {
+OnnxifiTransformer::OnnxifiTransformer(bool infer_shapes, bool debug)
```
Probably consider creating a "Config" struct to store these configuration parameters, othrewise everytime we add new one will need to touch the whole chain of function calls.

yinghai replied:
Yes. My tolerance is 3-4. :)

yinghai commented:
@bddppq Addressed comments.

Commit:
Enhance shape inference in ONNXIFI transformer (#12685)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/12685

In this diff, we push the fake run of the net into the ONNXIFI transformer, because
1. We cannot do shape inference for every op
2. Since the net has been SSA rewritten, we cannot use shape info from outer workspace directly.

In addition, this diff adds input shape info when querying the `onnxBackendCompatibility` function.

Differential Revision: D10390164

fbshipit-source-id: afb48738f790858dc33255bd3ea0065f3b2797d8
